### PR TITLE
perf(tokio): add improved runtime thread handling - multi cpu 

### PR DESCRIPTION
**Pull request by j-mendez**
* improve multi cpu architecture performance
* add improved thread runtime performance

-- note

This PR leverages multi-core CPUs and improves the spawn capabilities for operations.
By default threads live for 10 seconds after idle which makes sense for re-using the same thread except
for applications that can be ran once and exited - non http servers etc.

## benchmarks

```
  Model Name:	MacBook Pro
  Model Identifier:	MacBookPro18,2
  Chip:	Apple M1 Max
  Total Number of Cores:	10 (8 performance and 2 efficiency)
  Memory:	64 GB
```

cmd: `deno run -A --unstable cli/bench/write_file.js`

Before:

```
time 346 ms rate 28901
time 326 ms rate 30674
time 319 ms rate 31347
time 316 ms rate 31645
time 316 ms rate 31645
time 317 ms rate 31545
time 315 ms rate 31746
time 315 ms rate 31746
time 315 ms rate 31746
time 314 ms rate 31847
time 316 ms rate 31645
time 314 ms rate 31847
time 314 ms rate 31847
time 315 ms rate 31746
time 316 ms rate 31645
time 317 ms rate 31545
time 314 ms rate 31847
time 318 ms rate 31446
time 314 ms rate 31847
time 318 ms rate 31446
time 319 ms rate 31347
time 337 ms rate 29673
time 340 ms rate 29411
time 340 ms rate 29411
time 343 ms rate 29154
time 328 ms rate 30487
time 325 ms rate 30769
time 315 ms rate 31746
time 320 ms rate 31250
time 316 ms rate 31645
time 319 ms rate 31347
time 318 ms rate 31446
time 330 ms rate 30303
time 323 ms rate 30959
time 322 ms rate 31055
time 351 ms rate 28490
time 353 ms rate 28328
time 333 ms rate 30030
time 348 ms rate 28735
time 315 ms rate 31746
time 316 ms rate 31645
time 344 ms rate 29069
time 315 ms rate 31746
time 315 ms rate 31746
time 314 ms rate 31847
time 315 ms rate 31746
time 319 ms rate 31347
time 338 ms rate 29585
time 331 ms rate 30211
time 318 ms rate 31446
```
---
After:

```
time 331 ms rate 30211
time 318 ms rate 31446
time 315 ms rate 31746
time 317 ms rate 31545
time 317 ms rate 31545
time 321 ms rate 31152
time 320 ms rate 31250
time 317 ms rate 31545
time 319 ms rate 31347
time 321 ms rate 31152
time 315 ms rate 31746
time 317 ms rate 31545
time 315 ms rate 31746
time 314 ms rate 31847
time 314 ms rate 31847
time 315 ms rate 31746
time 316 ms rate 31645
time 315 ms rate 31746
time 314 ms rate 31847
time 316 ms rate 31645
time 316 ms rate 31645
time 315 ms rate 31746
time 315 ms rate 31746
time 315 ms rate 31746
time 315 ms rate 31746
time 314 ms rate 31847
time 315 ms rate 31746
time 315 ms rate 31746
time 315 ms rate 31746
time 315 ms rate 31746
time 315 ms rate 31746
time 315 ms rate 31746
time 317 ms rate 31545
time 313 ms rate 31948
time 315 ms rate 31746
time 316 ms rate 31645
time 315 ms rate 31746
time 316 ms rate 31645
time 315 ms rate 31746
time 315 ms rate 31746
time 315 ms rate 31746
time 314 ms rate 31847
time 314 ms rate 31847
time 310 ms rate 32258
time 317 ms rate 31545
time 316 ms rate 31645
time 314 ms rate 31847
time 315 ms rate 31746
time 317 ms rate 31545
time 314 ms rate 31847
```

### CODE REVIEW

File: runtime/tokio_util.rs [hunk 1]
```diff
@@ -2,14 +2,8 @@
...
   tokio::runtime::Builder::new_current_thread()
-    .enable_io()
-    .enable_time()
-    // This limits the number of threads for blocking operations (like for
-    // synchronous fs ops) or CPU bound tasks like when we run dprint in
-    // parallel for deno fmt.
-    // The default value is 512, which is an unhelpfully large thread pool. We
-    // don't ever want to have more than a couple dozen threads.
-    .max_blocking_threads(32)
+    .enable_all()
```
Comment (littledivy) [inline runtime/tokio_util.rs:5]: Let's keep it explicit with `enable_io` and `enable_time`.

**Comment by ry** [comment]
Thanks for the patch! Do you have some benchmark that demonstrates the improvement?

**Comment by j-mendez** [comment]
@ry you are welcome! I can get a benchmark this week. Need to make a minimal demo since the project is private, should be straight forward for the example. Thank you for the JS runtime using Rust!

**Comment by j-mendez** [comment]
> @ry you are welcome! I can get a benchmark this week. Need to make a minimal demo since the project is private, should be straight forward for the example. Thank you for the JS runtime using Rust!

Busy week - made a little progress on the benchmarks will try to get it done soon. Feel free to take this into other hands if moving too slow!

**Comment by j-mendez** [comment]
@ry pardon the delay, I went ahead and attached benchmarks for the runtime changes. Thank you for the patience.

**Comment by littledivy** [comment]
@j-mendez Thank you for the benchmarks. Although none of them are doing anything asynchrnous so it's unlikely that we are measuring `spawn` / `spawn_blocking` performance here. A better benchmark would be `Deno.writeFile` or `Deno.readFile` which uses `spawn_blocking` for file read/writes.

**Comment by j-mendez** [comment]
> @j-mendez Thank you for the benchmarks. Although none of them are doing anything asynchrnous so it's unlikely that we are measuring `spawn` / `spawn_blocking` performance here. A better benchmark would be `Deno.writeFile` or `Deno.readFile` which uses `spawn_blocking` for file read/writes.

You are welcome, I appreciate the lead on the next steps. I am going to see if I can look into that since running the bench at root is a bit harder to track due to the amount of benches ran.

**Comment by littledivy** [comment]
@j-mendez Try this one `deno run -A --unstable cli/bench/write_file.js`. It tries to write small chunks to `/dev/null` as fast as possible.

**Comment by j-mendez** [comment]
> @j-mendez Try this one `deno run -A --unstable cli/bench/write_file.js`. It tries to write small chunks to `/dev/null` as fast as possible.

I attached the logs of the before and after. I noticed afterwards my `rust-analyzer` ran in between reason for the drastic differences on branch checkouts.

**Comment by littledivy** [comment]
Ran the microbenchmarks again and i see no difference. I'd prefer to stay to the default unless there is a clear perf improvement.

**Pull request closed without merging**